jsonpb/jsonpb.go

```diff
@@ -528,6 +553,9 @@ func (u *Unmarshaler) UnmarshalNext(dec *json.Decoder, pb proto.Message) error {
 	if err := dec.Decode(&inputValue); err != nil {
 		return err
 	}
+	if jsu, ok := pb.(JSONPBUnmarshaler); ok {
```

Comment:
How about moving this inside unmarshalValue instead?  And I think it may need to be after block that allocates memory for pointer fields, i.e. 597-600.

You can do type assertion using ...
if jsu, ok := target.Addr().Interface().(JSONPBUnmarshaler); ok { ... }